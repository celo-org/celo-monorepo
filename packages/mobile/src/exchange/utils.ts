import BigNumber from 'bignumber.js'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { convertToContractDecimals } from 'src/tokens/saga'
import { getContractKitAsync } from 'src/web3/contracts'

export async function celoTocUsd(celoAmount: BigNumber) {
  const celoAmountInWei = (
    await convertToContractDecimals(celoAmount, CURRENCY_ENUM.GOLD)
  ).integerValue()

  const contractKit = await getContractKitAsync()
  const exchange = await contractKit.contracts.getExchange()
  const exchangeRate = await exchange.getGoldExchangeRate(celoAmountInWei)
  const cUsdAmount = exchangeRate.multipliedBy(celoAmount)

  return cUsdAmount
}
