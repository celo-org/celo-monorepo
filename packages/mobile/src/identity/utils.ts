import BigNumber from 'bignumber.js'

export function isUserBalanceSufficient(
  userBalance: string | number | null,
  estimatedTxAmount: number | string
) {
  if (!userBalance || new BigNumber(userBalance).isLessThan(new BigNumber(estimatedTxAmount))) {
    return false
  }

  return true
}
