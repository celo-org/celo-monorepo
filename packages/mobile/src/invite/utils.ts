import { sanitizeMessageBase64 } from '@celo/utils/src/attestations'
import URLSearchParamsReal from '@ungap/url-search-params'
import { Platform } from 'react-native'
import RNInstallReferrer from 'react-native-install-referrer'
import Logger from 'src/utils/Logger'

export const createInviteCode = (privateKey: string) => {
  // TODO(Rossy) we need some scheme to encrypt this PK
  // Buffer.from doesn't expect a 0x for hex input
  const privateKeyHex = privateKey.substring(2)
  return Buffer.from(privateKeyHex, 'hex').toString('base64')
}

// exported for testing
export const extractInviteCode = (inviteFieldInput: string) => {
  const sanitizedCode = sanitizeMessageBase64(inviteFieldInput)
  const regex = new RegExp('([0-9A-Za-z/\\+\\-\\_]*=)')
  const matches = sanitizedCode.match(regex)
  if (matches == null || matches.length === 0) {
    return null
  }
  return '0x' + Buffer.from(matches[0], 'base64').toString('hex')
}

// TODO(cmcewen): Consider web3 utils
export const isValidPrivateKey = (hexEncodedPrivateKey: string): boolean => {
  // First two chars are 0x
  if (hexEncodedPrivateKey.length !== 64 + 2) {
    return false
  }
  if (!hexEncodedPrivateKey.startsWith('0x')) {
    return false
  }
  return true
}

export function extractValidInviteCode(inviteFieldInput: string) {
  const inviteCode = extractInviteCode(inviteFieldInput)
  if (inviteCode == null || !isValidPrivateKey(inviteCode)) {
    return null
  } else {
    return inviteCode
  }
}

interface ReferrerData {
  clickTimestamp: string
  installReferrer: string
  installTimestamp: string
}

interface ReferrerDataError {
  message: string
}

export const getValidInviteCodeFromReferrerData = async () => {
  if (Platform.OS === 'android') {
    const referrerData: ReferrerData | ReferrerDataError = await RNInstallReferrer.getReferrer()
    Logger.info(
      'invite/utils/getInviteCodeFromReferrerData',
      'Referrer Data: ' + JSON.stringify(referrerData)
    )
    if (referrerData && referrerData.hasOwnProperty('installReferrer')) {
      const params = new URLSearchParamsReal(
        decodeURIComponent((referrerData as ReferrerData).installReferrer)
      )
      const inviteCode = params.get('invite-code')
      if (inviteCode) {
        const sanitizedCode = inviteCode.replace(' ', '+')
        // Accept invite codes which are either base64 encoded or direct hex keys
        if (isValidPrivateKey(sanitizedCode)) {
          return sanitizedCode
        }
        return extractValidInviteCode(sanitizedCode)
      }
    }
  }
  return null
}
