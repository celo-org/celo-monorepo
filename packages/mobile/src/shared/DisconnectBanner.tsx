import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { StyleSheet, Text } from 'react-native'
import { connect } from 'react-redux'
import componentWithAnalytics from 'src/analytics/wrapper'
import { Namespaces, withTranslation } from 'src/i18n'
import { RootState } from 'src/redux/reducers'
import { isAppConnected, isAppSynced } from 'src/redux/selectors'

interface StateProps {
  fornoEnabled: boolean
  appConnected: boolean
  appSynced: boolean
}

type Props = StateProps & WithTranslation

const mapStateToProps = (state: RootState): StateProps => {
  return {
    fornoEnabled: state.web3.fornoMode,
    appConnected: isAppConnected(state),
    appSynced: isAppSynced(state),
  }
}

class DisconnectBanner extends React.PureComponent<Props> {
  // This component is used in many screens but needs to remember when the app  been conneted.
  // This flag tracks that. Could move to redux but no need yet as it's the only consumer
  static hasAppConnected = false
  static hasAppSynced = false

  componentDidUpdate() {
    if (this.props.appConnected && !DisconnectBanner.hasAppConnected) {
      DisconnectBanner.hasAppConnected = true
    }
    if (this.props.appSynced && !DisconnectBanner.hasAppSynced) {
      DisconnectBanner.hasAppSynced = true
    }
  }

  render() {
    const { t, appConnected, appSynced, fornoEnabled } = this.props
    const appSyncedIfNecessary = appSynced || fornoEnabled

    // App's connected: show nothing
    if (appConnected && appSyncedIfNecessary) {
      return null
    }

    // App's connected, was synced, and now resyncing to new blocks: show nothing
    if (appConnected && !appSyncedIfNecessary && DisconnectBanner.hasAppSynced) {
      return null
    }

    // App's not connected but was before: show red banner
    if (!appConnected && DisconnectBanner.hasAppConnected) {
      return (
        <Text style={[styles.text, styles.textRed]}>
          <Text style={fontStyles.bold}>{t('poorConnection.0')}</Text> {t('poorConnection.1')}
        </Text>
      )
    }

    // App is connecting for first time, show grey banner
    return (
      <Text style={[styles.text, styles.textGrey, fontStyles.bold]} testID="connectingToCeloBanner">
        {t('connectingToCelo')}
      </Text>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    opacity: 0.5,
  },
  text: {
    ...fontStyles.bodySmall,
    textAlign: 'center',
    // Unset explicit lineHeight set by fonts.tsx otherwise the text is not centered vertically
    lineHeight: undefined,
  },
  textGrey: {
    color: colors.disconnectBannerGrey,
  },
  textRed: {
    color: colors.disconnectBannerRed,
  },
})

export default componentWithAnalytics(
  connect<StateProps, {}, {}, RootState>(mapStateToProps)(
    withTranslation(Namespaces.global)(DisconnectBanner)
  )
)
