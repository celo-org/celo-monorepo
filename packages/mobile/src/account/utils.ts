import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
