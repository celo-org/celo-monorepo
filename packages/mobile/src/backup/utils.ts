import { sampleSize } from 'lodash'
import { AsyncStorage } from 'react-native'
import { generateMnemonic, wordlists } from 'react-native-bip39'
import { getKey, setKey } from 'src/utils/keyStore'
import Logger from 'src/utils/Logger'

const TAG = 'Backup/utils'

export async function createQuizWordList(mnemonic: string, language: string | null) {
  const disallowedWordSet = new Set(mnemonic.split(' '))
  const languageWordList = getWordlist(language)
  const wordOptions: string = await generateMnemonic(1000, null, languageWordList)
  const quizWordList = new Set(
    [...wordOptions.split(' ')].filter((word: string) => !disallowedWordSet.has(word))
  )
  return [...quizWordList]
}

export function selectQuizWordOptions(correctWord: string, allWords: string[], numOptions: number) {
  const wordOptions = []
  const correctWordPosition = Math.floor(Math.random() * numOptions)
  const randomWordIndexList = sampleSize([...Array(allWords.length).keys()], numOptions - 1)
  let randomWordIndex: number = 0

  for (let i = 0; i < numOptions; i++) {
    if (i === correctWordPosition) {
      wordOptions.push(correctWord)
      continue
    }

    wordOptions.push(allWords[randomWordIndexList[randomWordIndex]])
    randomWordIndex += 1
  }
  return wordOptions
}

export function getWordlist(language: string | null) {
  if (!language) {
    return wordlists.EN
  }

  switch (language.slice(0, 2)) {
    case 'es': {
      return wordlists.ES
    }
    default: {
      return wordlists.EN
    }
  }
}

// TODO(Rossy) Remove after the next alfa testnet reset
export async function getStoredMnemonic(): Promise<string | null> {
  try {
    Logger.debug(TAG, 'Checking keystore for mnemonic')
    let mnemonic = await getKey('mnemonic')
    if (mnemonic) {
      return mnemonic
    }

    Logger.debug(TAG, 'Mnemonic not found in keystore, checking async storage')
    mnemonic = await AsyncStorage.getItem('mnemonic')
    if (mnemonic) {
      await setKey('mnemonic', mnemonic)
      await AsyncStorage.removeItem('mnemonic')
      return mnemonic
    }

    Logger.error(TAG, 'No mnemonic found')
    return null
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}
