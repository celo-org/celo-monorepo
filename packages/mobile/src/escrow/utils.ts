import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}
