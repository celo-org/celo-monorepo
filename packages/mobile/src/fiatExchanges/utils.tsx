import { SIMPLEX_URI } from 'src/config'
import { CURRENCY_ENUM } from 'src/geth/consts'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { navigateToURI } from 'src/utils/linking'

export const openMoonpay = (currencyCode: LocalCurrencyCode, currencyToBuy: CURRENCY_ENUM) => {
  navigate(Screens.MoonPay, {
    localAmount: 0,
    currencyCode,
    currencyToBuy,
  })
}

export const openSimplex = (account: string | null) => {
  navigateToURI(`${SIMPLEX_URI}?address=${account}`)
}
