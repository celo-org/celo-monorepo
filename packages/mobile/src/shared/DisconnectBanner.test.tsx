import * as React from 'react'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import DisconnectBanner from 'src/shared/DisconnectBanner'
import { createMockStore, createMockStoreAppDisconnected } from 'test/utils'

it('renders banner when app is disconnected', () => {
  const store = createMockStoreAppDisconnected()
  const tree = renderer.create(
    <Provider store={store}>
      <DisconnectBanner />
    </Provider>
  )
  expect(tree).toMatchSnapshot()
})

it('renders nothing when connected', () => {
  const store = createMockStore()
  const tree = renderer.create(
    <Provider store={store}>
      <DisconnectBanner />
    </Provider>
  )
  expect(tree.toJSON()).toBeNull()
})
