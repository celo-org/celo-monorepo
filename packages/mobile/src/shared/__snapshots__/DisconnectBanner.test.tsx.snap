// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders banner when app is disconnected 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#2E3338",
        "fontFamily": "Hind-Regular",
        "fontSize": 14,
        "lineHeight": undefined,
        "textAlign": "center",
      },
      Object {
        "color": "rgba(46, 51, 56, 0.4)",
      },
      Object {
        "fontFamily": "Hind-Bold",
      },
    ]
  }
  testID="connectingToCeloBanner"
>
  connectingToCelo
</Text>
`;
